### PR TITLE
Updated CHIPCert APIs to Use ByteSpans

### PR DESCRIPTION
**Pull request by emargolis**
#### Problem
CHIPCert APIs: not all of them use ByteSpan for input/output params and they don't integrate well with the rest of the code.  

#### Change overview
  -- Updated CHIPCert APIs to Use ByteSpans
  -- added new method TLVReader::Init() that initializes reader with ByteSpan.
  -- using ArraySize() where appropriate
  -- Other minor cleanups

#### Testing
unit testing

### CODE REVIEW

**Comment by bzbarsky-apple** [comment]
>  -- added new method TLVReader::Init() that initializes reader with ByteSpan.

Fwiw, that probably overlaps with my changes in https://github.com/project-chip/connectedhomeip/pull/8979

File: src/lib/core/CHIPTLVReader.cpp
Comment (bzbarsky-apple) [inline src/lib/core/CHIPTLVReader.cpp:59]: Why is this invalid?  Reading an empty thing might be reasonable, depending on what the caller is doing....  The actual reads will fail as needed.

File: src/lib/core/CHIPTLVReader.cpp
Comment (bzbarsky-apple) [inline src/lib/core/CHIPTLVReader.cpp:61]: Two-arg Init() now takes `size_t`, so no need for the cast bits.

Comment (emargolis) [reply src/lib/core/CHIPTLVReader.cpp]: updated

File: src/lib/core/CHIPTLVReader.cpp
Comment (emargolis) [reply src/lib/core/CHIPTLVReader.cpp]: updated

**Comment by emargolis** [comment]
> > -- added new method TLVReader::Init() that initializes reader with ByteSpan.
> 
> Fwiw, that probably overlaps with my changes in #8979

I reverted my changes in the TLVReader. Will probably wait for #8979 to land first and then will do some minor updates in this PR.

File: src/credentials/CHIPCert.h
Comment (tcarmelveilleux) [inline src/credentials/CHIPCert.h:627]: A bytespan is already const

**Review by tcarmelveilleux — APPROVED** [review]

File: src/credentials/CHIPCert.h
Comment (bzbarsky-apple) [reply src/credentials/CHIPCert.h:627]: A ByteSpan is a span over const bytes, but the object itself is not const.

That said, passing it by value means that mutating it (e.g. changing the size, or reassigning it to point to a different buffer altogether) won't affect the caller anyway... the benefit of making it const at that point is to make it clear that you should not do these things.  That said, if it's const anyway, worth checking whether const by value or const ref leads to smaller code or whether it's all the same in the end.

**Review by andy31415 — APPROVED** [review]

File: src/controller/ExampleOperationalCredentialsIssuer.cpp
Comment (bzbarsky-apple) [inline src/controller/ExampleOperationalCredentialsIssuer.cpp:113]:
```suggestion
    reader.Init(csrElements);
```

File: src/credentials/CHIPCert.cpp
Comment (bzbarsky-apple) [inline src/credentials/CHIPCert.cpp:139]:
```suggestion
    reader.Init(chipCert);
```

File: src/credentials/CHIPCert.cpp
Comment (bzbarsky-apple) [inline src/credentials/CHIPCert.cpp:242]:
```suggestion
    reader.Init(chipCert);
```

File: src/credentials/CHIPCertFromX509.cpp
Comment (bzbarsky-apple) [inline src/credentials/CHIPCertFromX509.cpp:719]:
```suggestion
    writer.Init(chipCert;
```

File: src/credentials/CHIPCertFromX509.cpp
Comment (bzbarsky-apple) [inline src/credentials/CHIPCertFromX509.cpp:782]:
```suggestion
    reader.Init(opCertArray);
```

File: src/credentials/CHIPCertToX509.cpp
Comment (bzbarsky-apple) [inline src/credentials/CHIPCertToX509.cpp:843]:
```suggestion
    reader.Init(chipCert);
```

File: src/credentials/CHIPCertToX509.cpp
Comment (bzbarsky-apple) [inline src/credentials/CHIPCertToX509.cpp:860]:
```suggestion
    reader.Init(chipCert);
```

**Review by bzbarsky-apple — APPROVED** [review]
Approving so this is not blocked on me, but the bug in CHIPDeviceController.cpp definitely needs to be fixed.

File: src/controller/ExampleOperationalCredentialsIssuer.cpp
Comment (emargolis) [reply src/controller/ExampleOperationalCredentialsIssuer.cpp]: the conflict is resolved.

File: src/controller/CHIPDeviceController.cpp
Comment (emargolis) [reply src/controller/CHIPDeviceController.cpp]: that was fixed.

**Review by woody-apple — APPROVED** [review]

File: src/credentials/CHIPCert.h
Comment (woody-apple) [reply src/credentials/CHIPCert.h:628]: @tcarmelveilleux ?

**Comment by andy31415** [comment]
@chrisdecenzo @jelderton @jmartinez-silabs @LuDuda @sagar-apple @msandstedt ?

**Review by msandstedt — APPROVED** [review]

**Pull request merged**
